src/haz3lcore/dynamics/TypeAssignment.re

```diff
@@ -0,0 +1,301 @@
+      | NotEquals => Some(Typ.Bool)
+      };
+    } else {
+      None;
+    };
+  | BinFloatOp(op, d1, d2) =>
+    let* ty1 = typ_of_dhexp(ctx, m, d1);
+    let* ty2 = typ_of_dhexp(ctx, m, d2);
+    if (Typ.eq(ty1, Float) && Typ.eq(ty2, Float)) {
+      switch (op) {
+      | Minus
+      | Plus
+      | Times
+      | Power
+      | Divide => Some(Typ.Float)
+      | LessThan
+      | LessThanOrEqual
+      | GreaterThan
+      | GreaterThanOrEqual
+      | Equals
+      | NotEquals => Some(Typ.Bool)
+      };
+    } else {
+      None;
+    };
+  | BinStringOp(op, d1, d2) =>
+    let* ty1 = typ_of_dhexp(ctx, m, d1);
+    let* ty2 = typ_of_dhexp(ctx, m, d2);
+    if (Typ.eq(ty1, String) && Typ.eq(ty2, String)) {
+      switch (op) {
+      | Concat => Some(Typ.String)
+      | Equals => Some(Typ.Bool)
+      };
+    } else {
+      None;
+    };
+  | ListLit(_, _, ty, _) => Some(List(ty))
+  | Cons(d1, d2) =>
+    let* ty1 = typ_of_dhexp(ctx, m, d1);
+    let* ty2 = typ_of_dhexp(ctx, m, d2);
+    switch (ty2) {
+    | List(ty3) when Typ.eq(ty3, ty1) => Some(ty2)
+    | _ => None
+    };
+  | ListConcat(d1, d2) =>
+    let* ty1 = typ_of_dhexp(ctx, m, d1);
+    let* ty2 = typ_of_dhexp(ctx, m, d2);
+    switch (ty1, ty2) {
+    | (List(ty1), List(ty2)) when Typ.eq(ty1, ty2) => Some(Typ.List(ty1))
+    | _ => None
+    };
+  | Tuple(dhs) =>
+    let+ typ_list =
+      dhs |> List.map(typ_of_dhexp(ctx, m)) |> OptUtil.sequence;
+    Typ.Prod(typ_list);
+  | Prj(dh, i) =>
+    let* ty = typ_of_dhexp(ctx, m, dh);
+    switch (ty) {
+    | Prod(l) when List.length(l) > i => Some(List.nth(l, i))
+    | _ => None
+    };
+  | Constructor(_, typ) => Some(typ)
+  | ConsistentCase(Case(d_scrut, d_rules, _)) =>
+    let* ty' = typ_of_dhexp(ctx, m, d_scrut);
+    let* typ_cases: list(Typ.t) =
+      d_rules
+      |> List.map((DHExp.Rule(dhp, de)) => {
+           let* ctx = dhpat_extend_ctx(dhp, ty', ctx);
+           typ_of_dhexp(ctx, m, de);
+         })
+      |> OptUtil.sequence;
+    equal_typ_list(typ_cases);
+  | Cast(d, ty1, ty2) =>
+    let* _ = Typ.join(~fix=true, ctx, ty1, ty2);
+    let* tyd = typ_of_dhexp(ctx, m, d);
+    Typ.eq(tyd, ty1) ? Some(ty2) : None;
+  | FailedCast(d, ty1, ty2) =>
+    if (ground(ty1) && ground(ty2) && !Typ.eq(ty1, ty2)) {
+      let* tyd = typ_of_dhexp(ctx, m, d);
+      Typ.eq(tyd, ty1) ? Some(ty2) : None;
+    } else {
+      None;
+    }
+  | InvalidOperation(d, _) => typ_of_dhexp(ctx, m, d)
+  | IfThenElse(ConsistentIf, d_scrut, d1, d2) =>
+    let* ty = typ_of_dhexp(ctx, m, d_scrut);
+    if (Typ.eq(ty, Bool)) {
+      let* ty1 = typ_of_dhexp(ctx, m, d1);
+      let* ty2 = typ_of_dhexp(ctx, m, d2);
+      equal_typ_list([ty1, ty2]);
+    } else {
+      None;
+    };
+  | IfThenElse(InconsistentIf, d_scrut, d1, d2) =>
+    let* ty = typ_of_dhexp(ctx, m, d_scrut);
+    if (Typ.eq(ty, Bool)) {
+      let* _ = typ_of_dhexp(ctx, m, d1);
+      let* _ = typ_of_dhexp(ctx, m, d2);
+      Some(Typ.Unknown(Internal));
```

Comment:
also here -- let's actually check to make sure that there is at least one branch with an inconsistent type